File: Maths/count_subgrid.cxx
/* Count the number of subgrid in nxn array ~O(n³) complexity
 * 
 * Source: Competitive Programmer’s Handbook (Antti Laaksonen) */

#include <bits/stdc++.h>
using namespace std;

class Solution{
	public:
	static int count_subgrid(vector<vector<int>> tab){
		int subgrids = 0;
		int n = tab.size();
		for(auto x = 0; x < n; x++){
			for(auto y = x+1; y < n; y++){ // loop over all pairs of (x,y) of rows
				int count = 0;
				for(auto  i = 0; i < n; i++){ // loop over all columns
					if((tab[x][i]==1)&(tab[y][i]==1))
						++count;
				}
				subgrids += ((count-1)*count)/2;
			}
		}return subgrids;
	}
};
int main(){
	ios_base::sync_with_stdio(0);
	vector<vector<int>> test = {{0,1,0,0,1},
								{0,1,1,0,0},
								{1,0,0,0,0},
								{0,1,1,0,1},
								{0,0,0,0,0}};
	cout << Solution::count_subgrid(test); // 2
}

File: Maths/count_subgrid.py
class Solution(object):
    def __init__(self,tab):
        self.tab = tab
    def count_subgrid(self,tab)->int:
        self.subgrids = 0
        self.n = len(tab)
        for x in range(self.n):
            for y in range(x+1,self.n):
                self.count = 0
                for i in range(self.n):
                    if((self.tab[x][i]==1)and(self.tab[y][i]==1)):
                        self.count+=1
                self.subgrids += ((self.count-1)*self.count)/2
        return self.subgrids

    

if __name__=='__main__':
    tab = [[0,1,0,0,1],
           [0,1,1,0,0],
           [1,0,0,0,0],
           [0,1,1,0,1],
           [0,0,0,0,0]]
    s = Solution(tab)
    print(s.count_subgrid(tab))
